users/vieting/experiments/librispeech/librispeech_100_ctc/fairseq_finetuning/ctc_standalone/pytorch_networks/wav2vec/w2v_ctc_wrapper.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+import torch
+from torch import nn
+
+import fairseq
+from fairseq.dataclass import FairseqDataclass
+from fairseq.models.wav2vec.wav2vec2_asr import Wav2VecCtc, Wav2Vec2CtcConfig
+
+from .w2v_ctc_wrapper_cfg import ModelConfig
+
+class Model(torch.nn.Module):
+    def __init__(self, w2v_config_updates, **kwargs):
+        """
+        Args:
+            w2v_config_updates: dictionary containing the configuration updates for the default FairseqConfig
+        """
+        super().__init__()
+        self.model_config = ModelConfig.from_dict(w2v_config_updates)
+        self.fairseq_config = self.model_config.build_full_config()
+        self.num_update = 0
+        dummy_task = self.model_config.build_dummy_task() # dummy task to build the model
+        self.w2v_model = Wav2VecCtc.build_model(self.fairseq_config.model, task=dummy_task) 
+
+    def get_num_updates(self):
+        """
+        Get the number of gradient updates that have been performed so far
+        """
+        return self.num_update
+    
+    def set_num_updates(self, num_update):
+        """
+        Set the number of gradient updates that have been performed so far
+        """
+        self.num_update = num_update
+        self.w2v_model.w2v_encoder.set_num_updates(num_update)
+
+    def forward(
+        self, 
+        raw_audio,
+        raw_audio_len,
+        **kwargs
+    ):
+        """
+        Args:
+            raw_audio: [B, T', 1] tensor (B: batch size, T': raw audio length)
+            raw_audio_len: [B] tensor
+            corpus_key: str
+        Returns:
+            logprobs: [B, T, C] tensor (C: vocab size, T: audio feature length)
+            input_lengths: [B] tensor
```

Comment:
Isn't it rather output lengths?

Reply:
Good point, `output_lengths` would be a better fitting name.

Reply:
Yes, but it corresponds to the parameter `input_lengths` for the CTC loss. IMO it's better to stay consistent and just name it the same way, otherwise we would call CTC loss with `input_lengths=output_lengths`, which would be equally confusing.

Reply:
A better possible name could be `input_feature_lengths` or just `feature_lengths,` what do you think?

Reply:
But that is outside of this function. You can use whatever variable name where you call this function, but I don't think the outside usage should define how we call the variables in here. `feature_lengths` seems alright here though.